import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java

```diff
@@ -82,70 +93,107 @@ protected void store(
-            transcoder.transcode((transcoder1, dataset) -> outputStream);
-          }
-        });
-      }
+        String sopClassUID = request.getString(Tag.AffectedSOPClassUID);
+        String sopInstanceUID = request.getString(Tag.AffectedSOPInstanceUID);
+        String transferSyntax = presentationContext.getTransferSyntax();
 
-      processorList.add((inputStream, outputStream) -> {
-        destinationClient.stowRs(inputStream);
-      });
+        validateParam(sopClassUID, "AffectedSOPClassUID");
+        validateParam(sopInstanceUID, "AffectedSOPInstanceUID");
 
-      processStream(association.getApplicationEntity().getDevice().getExecutor(),
-          inWithHeader, processorList);
+        final CountingInputStream countingStream;
 
-      response.setInt(Tag.Status, VR.US, Status.Success);
-      MonitoringService.addEvent(Event.CSTORE_BYTES, countingStream.getCount());
-    } catch (DicomWebException e) {
-      reportError(e);
-      DicomServiceException serviceException = new DicomServiceException(e.getStatus(), e);
-      serviceException.setErrorComment(e.getMessage());
-      throw serviceException;
+        if(destinationMap != null){
+          DicomInputStream inDicomStream  = new DicomInputStream(inPdvStream);
+          inDicomStream.mark(Integer.MAX_VALUE);
+          Attributes attrs = inDicomStream.readDataset(-1, Tag.PixelData);
+          inDicomStream.reset();
+
+          countingStream = new CountingInputStream(inDicomStream);
+          destinationClient.set(selectDestinationClient(association.getAAssociateAC().getCallingAET(), attrs));
+        } else {
+          countingStream = new CountingInputStream(inPdvStream);
+          destinationClient.set(defaultDicomWebClient);
+        }
+
+        InputStream inWithHeader =
+            DicomStreamUtil.dicomStreamWithFileMetaHeader(
+                sopInstanceUID, sopClassUID, transferSyntax, countingStream);
+
+        List<StreamProcessor> processorList = new ArrayList<>();
+        if (redactor != null) {
+          processorList.add(redactor::redact);
+        }
+
+        if(transcodeToSyntax != null && !transcodeToSyntax.equals(transferSyntax)) {
+          processorList.add((inputStream, outputStream) -> {
+            try (Transcoder transcoder = new Transcoder(inputStream)) {
+              transcoder.setIncludeFileMetaInformation(true);
+              transcoder.setDestinationTransferSyntax(transcodeToSyntax);
+              transcoder.transcode((transcoder1, dataset) -> outputStream);
+            }
+          });
+        }
+
+        if (backupUploadService != null) {
+          processorList.add((inputStream, outputStream) -> {
+            byte[] bytes;
+            try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+              StreamUtils.copy(inputStream, baos);
+              bytes = baos.toByteArray();
+              backupState.set(backupUploadService.createBackup(bytes, sopInstanceUID));
```

Comment:
Buffering the entire file in memory here is going to be really costly for some DICOM files like MRIs (200MB each I've seen). Can we do a streaming copy to the file location here instead of buffering the entire file into memory? And then a streaming read from the backup location to the healtcare API?

Reply:
Done. But this approach adds new risks on situation when --persistent_file_storage_location is set -> BackupUploadService created: if local or GCP uploader fails on read / write to backup file, then file will not be uploaded. Because last processor in the CStoreService (https://github.com/NikolaiKovalenko/healthcare-dicom-dicomweb-adapter/blob/to_upstream_retries/import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java#L157) depends on the result of previous processor that failed, and it`s OutputStream will be empty or in invalid state.

Reply:
@danielbeaudreau About you question - "Did you reproduce this during testing?". Answer is - yes. Here the test:  https://github.com/NikolaiKovalenko/healthcare-dicom-dicomweb-adapter/blob/to_upstream_retries/import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreServiceTest.java#L314